crates/subspace-networking/src/shared.rs

```diff
@@ -79,3 +85,61 @@ impl Shared {
         }
     }
 }
+
+// Supports exact keys for Kademlia. It uses the Multihash as the canonical type
+// and reduces all the conversions to Multihash conversion.
+// It supports only Sr25519 key pairs for PeerId and Identity multihashes.
+#[derive(Clone, Debug)]
+pub struct IdendityHash(GenericArray<u8, U32>);
+
+impl PreimageIntoKeyBytes<Multihash> for IdendityHash {
+    fn preimage_into_key_bytes(multihash: &Multihash) -> KeyBytes<Self> {
+        const SHA256_HASH_SIZE: usize = 32;
+
+        // Identity hasher for 32-bytes digest
+        if multihash.code() == u64::from(Code::Identity)
+            && multihash.digest().len() == SHA256_HASH_SIZE
+        {
+            let array = GenericArray::from_slice(multihash.digest());
+            return KeyBytes::from_unchecked(*array);
+        }
+
+        panic!(
+            "Unsupported multihash type. Expected Identity:{:?}",
+            multihash
+        )
+    }
+}
+
+impl PreimageIntoKeyBytes<PeerId> for IdendityHash {
+    fn preimage_into_key_bytes(peer_id: &PeerId) -> KeyBytes<Self> {
+        let multihash: Multihash = if let Ok(public_key) =
+            identity::PublicKey::from_protobuf_encoding(peer_id.as_ref().digest())
+        {
+            if let identity::PublicKey::Sr25519(pk) = public_key {
+                Code::Identity.digest(&pk.encode())
+            } else {
+                panic!("Unsupported public key type. Expected Sr25519")
```

Comment:
Same here, it shouldn't panic as there is a chance it'll be sent by the network and crash an app using this library.

If you're certain that is impossible, provide a convincing explanation here before panicking.

Reply:
Changed to the default SHA256 converter with a code comment.